### PR TITLE
Add git-lfs to the image.

### LINKED ISSUE #4: Please add git-lfs support
The git that is shipped with this image doesn't understand git-lfs, and so when testing against a repository that uses lfs, it fails.

### PR DESCRIPTION
**Pull request by pwagland**
Fixes #4.

### CODE REVIEW

**Review by cptactionhank — COMMENTED** [review]
Isn't there a less verbose to add GPG keys?

**Comment by pwagland** [comment]
Potentially you could drop the key into a file, and import from the file. But that would add an extra layer into the resulting image.

**Comment by pwagland** [comment]
Let me know if you want that change made, and I'll update the pull request.

**Comment by pwagland** [comment]
I have update the request to split the key out into its own file, and updated the Dockerfile to follow best practices from https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/ . This means that the image ends up slightly smaller, even though git-lfs has been added.

**Comment by pwagland** [comment]
@cptactionhank Is there anything different that you would like to see to get this merged in?

**Comment by cptactionhank** [comment]
Yes, I have concerns about the practice of doing it this way and are examining how some of the other official images handles this eg. postgres, and others.

I'll get back to you

File: Dockerfile [hunk 1]
```diff
@@ -25,6 +37,12 @@ RUN set -x \
         --delete              "Server/Service/Engine/Host/@xmlValidation" \
         --delete              "Server/Service/Engine/Host/@xmlNamespaceAware" \
                               "${BAMBOO_INSTALL}/conf/server.xml" \
+    && echo "Removing packages that are no longer needed" \
+    && dpkg --purge \
+            apt-transport-https \
+            xmlstarlet \
```
Comment (cptactionhank) [inline Dockerfile]: `xmlstarlet` is used by the entrypoint script and shall not be purged

File: Dockerfile [hunk 2]
```diff
@@ -5,13 +5,25 @@ ENV BAMBOO_HOME     /var/atlassian/bamboo
...
+# Copy in the gpg key for github_git-lfs so that we can load git-lfs
+# into the image.
+COPY github-lfs-repo.gpgkey /tmp
+
 # Install Atlassian Bamboo and helper tools and setup initial home
 # directory structure.
 RUN set -x \
+    && echo 'Adding repository and required key to install git-lfs' \
+    && apt-key add /tmp/github-lfs-repo.gpgkey \
+    && rm /tmp/github-lfs-repo.gpgkey \
```
Comment (cptactionhank) [inline Dockerfile]: Please remove this. 

A file layer with the key still exists so the removal doesn't provide much value

File: Dockerfile [hunk 3]
```diff
@@ -5,13 +5,25 @@ ENV BAMBOO_HOME     /var/atlassian/bamboo
 ENV BAMBOO_INSTALL  /opt/atlassian/bamboo
 ENV BAMBOO_VERSION  6.0.0
 
+# Copy in the gpg key for github_git-lfs so that we can load git-lfs
+# into the image.
+COPY github-lfs-repo.gpgkey /tmp
```
Comment (cptactionhank) [inline Dockerfile]: Please remove these added lines

File: Dockerfile [hunk 4]
```diff
@@ -5,13 +5,25 @@ ENV BAMBOO_HOME     /var/atlassian/bamboo
...
+
 # Install Atlassian Bamboo and helper tools and setup initial home
 # directory structure.
 RUN set -x \
+    && echo 'Adding repository and required key to install git-lfs' \
```
Comment (cptactionhank) [inline Dockerfile]: Please remove this line

File: Dockerfile [hunk 5]
```diff
@@ -5,13 +5,25 @@ ENV BAMBOO_HOME     /var/atlassian/bamboo
...
+    && apt-key add /tmp/github-lfs-repo.gpgkey \
```
Comment (cptactionhank) [inline Dockerfile]: Please remove this (it could also be added using `curl` and a pipe command)

File: github-lfs-repo.gpgkey [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+bgmhqRHlgVeKn9pwN4cpVBN+idzwPefQksSKH4lBDvVr/9j+V9mmrOx7QmQ5LCc/
+1on+L0dqo6suoajADhKy+lDQbzs2mVb4CLpPKncDup/9iJbjiR17DDFMwgyCoy5O
+HJICQ5lckNNgkHTS6Xiogkt28YfK4P3S0GaZgIrhKQ7AmO3O+hB12Zr+olpeyhGB
+OpBD80URntdEcenvfnXBY/BsuAVbTGXiBzrlBEyQxg656jUeqAdXg+nzCvP0yJlB
+UOjEcwyhK/U2nw9nGyaR3u0a9r24LgijGpdGabIeJm6O9vuuqFHHGI72pWUEs355
+lt8q1pAoJUv8NehQmlaR0h5wcwhEtwM6fiSIUTnuJnyHT053GjsUD7ef5fY1KEFm
+aZeW04kRtFDOPinz0faE8hvsxzsVgkKye1c2vkXKdOXvA3x+pZzlTHtcgMOhjKQA
+sA==
+=H60S
+-----END PGP PUBLIC KEY BLOCK-----
```
Comment (cptactionhank) [inline github-lfs-repo.gpgkey]: Please remove this whole file

File: Dockerfile [hunk 7]
```diff
@@ -5,13 +5,25 @@ ENV BAMBOO_HOME     /var/atlassian/bamboo
...
+    && rm /tmp/github-lfs-repo.gpgkey \
+    && apt-get update --quiet \
+    && apt-get install --quiet --yes --no-install-recommends \
+               apt-transport-https \
+               xmlstarlet \
+    && echo 'deb http://packagecloud.io/github/git-lfs/debian/ jessie main' > /etc/apt/sources.list.d/github_git-lfs.list \
```
Comment (cptactionhank) [inline Dockerfile]: Please remove this

File: Dockerfile [hunk 8]
```diff
@@ -5,13 +5,25 @@ ENV BAMBOO_HOME     /var/atlassian/bamboo
...
     && apt-get update --quiet \
-    && apt-get install --quiet --yes --no-install-recommends xmlstarlet \
+    && apt-get install --quiet --yes --no-install-recommends \
+               git-lfs \
```
Comment (cptactionhank) [inline Dockerfile]: git-lfs/git-lfs has a docker recipe stated as

```
RUN build_deps="curl ca-certificates" && \
    apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends ${build_deps} && \
    curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | bash && \
    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends git-lfs && \
    git lfs install && \
    DEBIAN_FRONTEND=noninteractive apt-get purge -y --auto-remove ${build_deps} && \
    rm -r /var/lib/apt/lists/*
```

although `curl` is already available in the `openjdk` image and `ca-certificates` is available as well which should boil it down to:
```
RUN apt-get update && \
    curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | bash && \
    apt-get install -y --no-install-recommends git-lfs && \
    git lfs install && \
    rm -r /var/lib/apt/lists/*
```

Please add the following lines instead

```
&& curl --silent https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | bash \
&& apt-get install --quiet --yes --no-install-recommends git-lfs \
&& git lfs install \
```

If we are to trust their GPG key we might as well trust their install script.

File: Dockerfile [hunk 9]
```diff
@@ -5,13 +5,25 @@ ENV BAMBOO_HOME     /var/atlassian/bamboo
...
 # Install Atlassian Bamboo and helper tools and setup initial home
 # directory structure.
 RUN set -x \
+    && echo 'Adding repository and required key to install git-lfs' \
+    && apt-key add /tmp/github-lfs-repo.gpgkey \
+    && rm /tmp/github-lfs-repo.gpgkey \
+    && apt-get update --quiet \
+    && apt-get install --quiet --yes --no-install-recommends \
+               apt-transport-https \
+               xmlstarlet \
```
Comment (cptactionhank) [inline Dockerfile]: I would prefer dependencies in a single line

File: Dockerfile (hunk 3 above)
Comment (pwagland) [reply Dockerfile]: If you remove those lines then you don't have the key for the new repository.

File: Dockerfile (hunk 5 above)
Comment (pwagland) [reply Dockerfile]: I don't like just pulling in random files from the net without some verification that I am getting a verified version of it.

The other way to do it that I have seen is to pull in the key from the central key servers, so something like (taken from the sonar-scanner image):

    # pub   2048R/D26468DE 2015-05-25
    #       Key fingerprint = F118 2E81 C792 9289 21DB  CAB4 CFCA 4A29 D264 68DE
    # uid                  sonarsource_deployer (Sonarsource Deployer) <infra@sonarsource.com>
    # sub   2048R/06855C1D 2015-05-25
    && gpg --keyserver ha.pool.sks-keyservers.net --recv-keys F1182E81C792928921DBCAB4CFCA4A29D26468DE \

File: Dockerfile (hunk 9 above)
Comment (pwagland) [reply Dockerfile]: To be honest, so do I, this is just following the guidelines of the "best practices document". If you are happy with one line, so am I!

File: Dockerfile (hunk 7 above)
Comment (pwagland) [reply Dockerfile]: If this is removed, then git-lfs cannot be (easily) installed. apt-transport-https is needed for the git-lfs repo, as it is only exported over https, via a redirect.

File: Dockerfile (hunk 8 above)
Comment (pwagland) [reply Dockerfile]: FWIW, that script basically just does the lines that I added, I pulled out the essential elements of that script into this file. But I am happy to just call that shell script if you are.

File: github-lfs-repo.gpgkey (hunk 6 above)
Comment (pwagland) [reply github-lfs-repo.gpgkey]: Done.

File: Dockerfile (hunk 8 above)
Comment (pwagland) [reply Dockerfile]: Done.

**Comment by pwagland** [comment]
Requested changes are in.

File: Dockerfile (hunk 5 above)
Comment (cptactionhank) [reply Dockerfile]: I get that and their supplied script does the same thing as you did originally, but I think it is simpler and have less potential maintenance doing it this way. The trusted GPG key is taken from the same place as where the install script is located so this should be just as trusted as well.

But I would prefer the other way you suggested in long term, this is what the official `postgres` image does as well.

**Comment by cptactionhank** [comment]
Acceptance tests works on my machine. Thank you for your work!

**Pull request merged**